Adds league table endpoint

janakerman opened the pull request:
Resolves #49 

Note that in the league table I'm 'unsafely' accessing
properties on an object. However, since the GameColor enum
fromString method returns an optional we're checking and logging
as we might as well.

I'm trying not to be too defensive while programming here, we should
assume that once we're handling game data that's been downloaded,
it should be safe, so we shouldn't null check everything.

I'll add to our item validator in the following commit so we can be sure
that we are dealing with safe objects internally.

Opinions on this wanted!

janakerman committed:
Adds league table endpoint

Note that in the league table I'm 'unsafely' accessing
properties on an object. However, since the GameColor enum
fromString method returns an optional we're checking and logging
as we might as well.

I'm trying not to be too defensive while programming here, we should
assume that once we're handling game data that's been downloaded,
it should be safe, so we shouldn't null check everything.

I'll add to our item validator in the following commit so we can be sure
that we are dealing with safe objects internally.

Opinions on this wanted!

bjedrzejewski commented on src/main/java/com/chessmates/league/LeagueTable.groovy:
```diff
@@ -0,0 +1,70 @@
+
+    private final Map<String, LeagueRow> playerRows = new HashMap()
+
+    LeagueTable(Integer winPoints, Integer loosePoints) {
```
In chess, draw is usually considered 0.5 point, so that I think we need to go beyond Integer.

janakerman replied:
Happy to add - I wasn't too sure how draws were handled in chess, but more importantly I'm not sure how Lichess indicates a draw. Do you have an example of a drawn game? Perhaps a game you've drawn with someone?

bjedrzejewski replied:
This is a draw: https://en.lichess.org/n8JDtXt1zaiF

bjedrzejewski commented on src/main/java/com/chessmates/league/LeagueTable.groovy:
```diff
@@ -0,0 +1,70 @@
+    private static final Logger logger = org.slf4j.LoggerFactory.getLogger(LeagueTable)
+
+    final winPoints
+    final loosePoints
```
What are loosePoints? Can we add a comment here?

janakerman replied:
Loose points are the amount of points given for a loss. May not make sense for a normal points system (1 for win, 0 for loss) but since we're providing these values as a constructor you could conceivably give some value to a lost game too. Will add a comment.

bjedrzejewski commented on src/main/java/com/chessmates/league/LeagueTable.groovy:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /** Add games to the league table. **/
+    protected void add(List games) {
```
This is a detail, but maybe we can accept any collection of games? Does not have to be ordered as a lit.

janakerman replied:
Good suggestion!

bjedrzejewski commented on src/main/java/com/chessmates/league/LeagueTable.groovy:
```diff
@@ -0,0 +1,70 @@
+            final maybeWinColor = GameColor.fromString(game.winner)
+
+            if (!maybeWinColor.isPresent()) {
+                logger.error "Couldn't add game to league: no winner field"
```
I would also log for which game object it was not present.

bjedrzejewski commented on src/main/java/com/chessmates/league/LeagueTable.groovy:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /** Add games to the league table. **/
+    protected void add(List games) {
```
This needs to handle draws as well

janakerman replied:
As above.

bjedrzejewski commented on src/main/java/com/chessmates/league/LeagueTable.groovy:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /** Represents a single players results in the league. */
+    final class LeagueRow {
```
I would make it into a static inner class or just take it out of this class all together and make it a separate one.

janakerman replied:
The reason it's an inner class and not static is that the league rows are referring to the points values given to the league. Happy to pull this out but would require each row to be initialised with the number of points for a win/loss/draw/. I was hoping to avoid a situation whereby two different rows could be in a league with different point values. Not really a possibility now as the league create its own rows.

bjedrzejewski replied:
Ok, my bad I have overlooked that it refers to that. Let's keep it how it is.

bjedrzejewski commented on src/main/java/com/chessmates/league/SimpleLeagueTableService.groovy:
```diff
@@ -0,0 +1,32 @@
+@Service
+class SimpleLeagueTableService implements LeagueTableService {
+
+    private final Integer WIN_POINTS = 1
```
I think we can add the DRAW_POINTS = 0.5 here

bjedrzejewski reviewed (changes requested):
Hey, some changes needed, mainly around handling of draws.

janakerman committed:
Updates for review feedback

- Adds property documentation
- Updates interface to accept collection
- Improves logging

janakerman committed:
Updates league table to handle draws

Draws are set to 0.5 points by our service.

janakerman committed:
Adds position to row object

Now we handle draws, we can't rely on list index to
indicate table position.

janakerman commented:
@bjedrzejewski - opinions on the rest endpoint? I'm considering changing it to:
table/league

If we treat a table as a resource, with the potential for there to be multiple tables in the future:
table/cross

Thoughts?

bjedrzejewski committed:
Merge branch 'master' into league-table
